lib/sampler/storage/src/test/span-dao-test.js

```diff
@@ -0,0 +1,540 @@
+    });
+
+    it('returns multiple spans, when applicable', async () => {
+      const spans = await dao.findIncompleteSpans(1534168800001, 0, 3); // 2018-08-13 14:00:00.001
+      expect(spans.length).to.equal(2);
+      expect(spans[0].target).to.deep.equal(firstTarget);
+      expect(spans[1].target).to.deep.equal(secondTarget);
+    });
+
+    it('returns only spans that are not yet complete', async () => {
+      const spans = await dao.findIncompleteSpans(1534176000001, 0, 3); // 2018-08-13 16:00:00.001
+      expect(spans.length).to.equal(2);
+      expect(spans[0].target).to.deep.equal(firstTarget);
+      expect(spans[1].target).to.deep.equal(secondTarget);
+    });
+
+    it('returns no candiates, if there are no matching', async () => {
+      const spans = await dao.findIncompleteSpans(1534161600000, 0, 3); // 2018-08-13 12:00:00.000
+      expect(spans.length).to.equal(0);
+    });
+
+    it('returns spans according to the limit setting', async () => {
+      const spans = await dao.findIncompleteSpans(1534168800001, 0, 1); // 2018-08-13 14:00:00.001
+      expect(spans.length).to.equal(1);
+      expect(spans[0].target).to.deep.equal(firstTarget);
+    });
+
+    it('returns spans according to the offset setting', async () => {
+      const spans = await dao.findIncompleteSpans(1534168800001, 1, 3); // 2018-08-13 14:00:00.001
+      expect(spans.length).to.equal(1);
+      expect(spans[0].target).to.deep.equal(secondTarget);
+    });
+  });
+
+  describe('#findCompleteSpans', () => {
+    let firstTarget;
+    let secondTarget;
+    let thirdTarget;
+
+    beforeEach(async () => {
+      firstTarget = createRandomTarget();
+      const firstSpanPO = spanPageObject(dao, firstTarget);
+      await firstSpanPO.startSpan(startTimestamp, measures).verify();
+      const firstInterval = {
+        start: startTimestamp,
+        end: 1534161600000 // 2018-08-13 12:00:00
+      };
+      await firstSpanPO.updateSpanProcessedInterval(firstInterval, true, 1).verify();
+
+      secondTarget = createRandomTarget();
+      const secondSpanPO = spanPageObject(dao, secondTarget);
+      await secondSpanPO.startSpan(startTimestamp, measures).verify();
+      const secondInterval = {
+        start: startTimestamp,
+        end: 1534168800000 // 2018-08-13 14:00:00
+      };
+      await secondSpanPO.updateSpanProcessedInterval(secondInterval, true, 1).verify();
+
+      thirdTarget = createRandomTarget();
+      const thirdSpanPO = spanPageObject(dao, thirdTarget);
+      await thirdSpanPO.startSpan(startTimestamp, measures).verify();
+      const thirdInterval = {
+        start: startTimestamp,
+        end: 1534176000000 // 2018-08-13 16:00:00
+      };
+      await thirdSpanPO.updateSpanProcessedInterval(thirdInterval, false, 1).verify();
+    });
+
+    it('returns only spans that are sampled earlier than specified', async () => {
+      const spans = await dao.findCompleteSpans(1534161600001, 0, 3); // 2018-08-13 12:00:00.001
```

Comment:
I was also thinking the same, but it becomes even harder to know what parameters you need to pass. Ok, in the test it will be clear, but if you need to use the method, you need to ensure that you don't miss a parameter. Looking at the actual code will not help either, since the method definition would look as follows.

```js
async findCompleteSpans(opts) {
}
```

Then again, we could use destructuring.


```js
async findCompleteSpans({before, offset, limit}) {
}
```

I am not sure on this.

Reply:
It feels to me to be related to lack of types in NodeJS and poor auto-completion in the IDEs that would otherwise help in other languages.

Reply:
We agreed that we will leave it as is for now.